### PR TITLE
TensorMap: Add Erase Gradient and Gradient Tests

### PR DESCRIPTION
**Pull request by kttian**
Update gradients and gradient tests for the TensorMap feature to be more comprehensive.

Add gradient for the erase op.
Implement ZerosLike and BinaryAdd kernels, which will be used in certain gradients.
Add gradient tests to cover more cases, including gradients for more function combinations, replace, erase, and non-scalar values.
@mdanatg @saxenasaurabh @dynamicwebpaige

### CODE REVIEW

**Review by mdanatg — DISMISSED** [review]

**Review by mdanatg — DISMISSED** [review]

File: tensorflow/python/ops/map_ops.py [hunk 1]
```diff
@@ -66,10 +66,14 @@ def LookupGrad(op, dval):
...
   _, k, v = op.inputs
   key_grad = None
-  value_grad = control_flow_ops.cond(
-      tensor_map_has_key(dmap, k), lambda: tensor_map_lookup(dmap, k, v.dtype),
-      lambda: array_ops.zeros_like(v))
-  map_grad = control_flow_ops.cond(
-      tensor_map_has_key(dmap, k),
-      lambda: tensor_map_erase(dmap, k, v.dtype)[0], lambda: dmap)
+  (value_grad, map_grad) = control_flow_ops.cond(tensor_map_has_key(dmap, k),
+                                                 lambda: (tensor_map_lookup(dmap, k, v.dtype), tensor_map_erase(dmap, k, v.dtype)[0]),
```
Comment (saxenasaurabh) [inline tensorflow/python/ops/map_ops.py:70]: I missed this earlier but why do we need a dtype arg for `TensorMapErase`?

File: tensorflow/python/ops/map_ops.py [hunk 2]
```diff
@@ -66,10 +66,14 @@ def LookupGrad(op, dval):
...
+                                                 lambda: (array_ops.zeros_like(v), dmap))
   return map_grad, key_grad, value_grad
+
+@ops.RegisterGradient("TensorMapErase")
+def EraseGrad(op, dmap, dval):
+  _, k = op.inputs
+  key_grad = None
+  map_grad = tensor_map_insert(dmap, k, dval)
```
Comment (saxenasaurabh) [inline tensorflow/python/ops/map_ops.py:78]: Another alternative would be to not return the erased tensor from `TensorMapErase`. Then `EraseGrad` would simply return `dmap`.

@mdanatg what do you think?

File: tensorflow/core/kernels/map_kernels.h [hunk 3]
```diff
@@ -175,6 +176,34 @@ class TensorMapHasKey : public OpKernel {
   }
 };
 
+template <typename Device>
+Status TensorMapBinaryAdd(OpKernelContext* c, const TensorMap& a,
+                           const TensorMap& b, TensorMap* out) {
```
Comment (saxenasaurabh) [inline tensorflow/core/kernels/map_kernels.h:181]: Could you add a comment here?

File: tensorflow/core/kernels/map_kernels.h [hunk 4]
```diff
@@ -175,6 +176,34 @@ class TensorMapHasKey : public OpKernel {
...
+    }
+    else {
+      out->tensors().emplace(p.first, p.second);
+    }
+  }
+  return Status::OK();
+}
+
+template <typename Device>
+Status TensorMapZerosLike(OpKernelContext* c, const TensorMap& x, TensorMap* y) {
```
Comment (saxenasaurabh) [inline tensorflow/core/kernels/map_kernels.h:198]: Could we avoid creating the zeros here and just return an empty map instead?

File: tensorflow/python/ops/map_ops.py (hunk 2 above)
Comment (mdanatg) [reply tensorflow/python/ops/map_ops.py:78]: Yes, I think this makes a good reason to tilt the balance in favor of not returning the erased value.

File: tensorflow/core/kernels/map_kernels.h [hunk 5]
```diff
@@ -196,11 +197,7 @@ Status TensorMapBinaryAdd(OpKernelContext* c, const TensorMap& a,
 
 template <typename Device>
 Status TensorMapZerosLike(OpKernelContext* c, const TensorMap& x, TensorMap* y) {
-  for (const std::pair<TensorKey,Tensor>& p : x.tensors()) {
-    Tensor val;
-    TF_RETURN_IF_ERROR(ZerosLikeTensor<Device>(c, p.second, &val));
-    y->tensors().emplace(p.first, val);
-  }
+  // zeros like returns an empty map
```
Comment (mdanatg) [inline tensorflow/core/kernels/map_kernels.h:200]: Nice!

File: tensorflow/core/kernels/map_kernels.h [hunk 6]
```diff
@@ -175,6 +176,31 @@ class TensorMapHasKey : public OpKernel {
   }
 };
 
+template <typename Device>
+Status TensorMapBinaryAdd(OpKernelContext* c, const TensorMap& a,
+                          const TensorMap& b, TensorMap* out) {
+  // binary add returns a union of keys, values with keys in the intersection are added
```
Comment (mdanatg) [inline tensorflow/core/kernels/map_kernels.h:182]: Style nit (for future changes; to expedite the submission, there's no need to update this one, unless other changes will be needed) - we prefer code comments to be well-formed sentences, with capitalization. Punctuation can be omitted if the comment is very short (2-3 words).

**Review by mdanatg — DISMISSED** [review]

**Review by mdanatg — APPROVED** [review]

**Pull request merged**